homework/81955/hw1/Lexer.cpp

```diff
@@ -0,0 +1,526 @@
+    res += c;
+    nextSymbol();
+    while (!match(c) && !Helper::isEof(code[index]))
+    {
+        res += code[index];
+        nextSymbol();
+    }
+    if (code[index - 1] == c)
+    {
+        res += c;
+    }
+    state = true;
+    return res;
+}
+
+std::string Tokenizer::parseNumber()
+{
+    char c = code[index];
+    if (!Helper::isDigit(c))
+    {
+        state = false;
+        return "";
+    }
+
+    nextSymbol();
+    std::string res;
+    res += c;
+
+    if (Helper::isBinLiteral(code[index]))
+    {
+        res += code[index];
+        nextSymbol();
+        while (Helper::isBinDigit(code[index]) && !Helper::isEof(code[index]))
+        {
+            res += code[index];
+            nextSymbol();
+        }
+        if (Helper::isBinLiteral(code[index - 1]))
+        {
+            state = false;
+            return res;
+        }
+        state = true;
+        return res;
+    }
+
+    if (Helper::isHexLiteral(code[index]))
+    {
+        res += code[index];
+        nextSymbol();
+        while (Helper::isHexDigit(code[index]) && !Helper::isEof(code[index]))
+        {
+            res += code[index];
+            nextSymbol();
+        }
+        if (Helper::isHexLiteral(code[index - 1]))
+        {
+            state = false;
+            return res;
+        }
+        state = true;
+        return res;
+    }
+
+    if (Helper::isOctalLiteral(code[index]))
+    {
+        res += code[index];
+        nextSymbol();
+        while (Helper::isOctalDigit(code[index]) && !Helper::isEof(code[index]))
+        {
+            res += code[index];
+            nextSymbol();
+        }
+        if (Helper::isOctalLiteral(code[index - 1]))
+        {
+            state = false;
+            return res;
+        }
+        state = true;
+        return res;
+    }
+
+    while (Helper::isDigit(code[index]) && !Helper::isEof(code[index]))
+    {
+        res += code[index];
+        nextSymbol();
+    }
+
+    state = true;
+    return res;
+}
+
+std::string Tokenizer::parseKeyword()
+{
+    size_t oldIndex = index;
+    if (!Helper::isValidIdentifierStart(code[index]))
+    {
+        state = false;
+        return "";
+    }
+
+    std::string res;
+
+    while (Helper::isValidIdentifierLetter(code[index]))
+    {
+        res += code[index];
+        nextSymbol();
+    }
+
+    if (keywordTable.find(res) != keywordTable.end())
+    {
+        state = true;
+        return res;
+    }
+    state = false;
+    index = oldIndex;
+    return "";
+}
+
+std::string Tokenizer::parseIdentifier()
+{
+    if (!Helper::isValidIdentifierStart(code[index]))
+    {
+        state = false;
+        return "";
+    }
+
+    std::string res;
+
+    while (Helper::isValidIdentifierLetter(code[index]) && !Helper::isEof(code[index]))
+    {
+        res += code[index];
+        nextSymbol();
+    }
+
+    state = true;
+    return res;
+}
+
+Token Tokenizer::nextToken()
+{
+
+    std::string whitespace = parseWhitespace();
+    if (state)
```

Comment:
`std::optional<std::string>` can help you avoid the `state` and will make the code simpler.